util/conditions/experimental/getter.go

```diff
@@ -0,0 +1,205 @@
+// - Objects with metav1.condition in status.conditions
+//
+// In case the object does not have metav1.conditions, GetAll tries to read clusterv1.condition from status.conditions
+// and convert them to metav1.conditions.
+func GetAll(obj runtime.Object) ([]metav1.Condition, error) {
+	if obj == nil {
+		return nil, errors.New("obj cannot be nil")
+	}
+
+	switch obj.(type) {
+	case runtime.Unstructured:
+		return getFromUnstructuredObject(obj)
+	default:
+		return getFromTypedObject(obj)
+	}
+}
+
+func getFromUnstructuredObject(obj runtime.Object) ([]metav1.Condition, error) {
+	u, ok := obj.(runtime.Unstructured)
+	if !ok {
+		// NOTE: this should not happen due to the type assertion before calling this fun
+		return nil, errors.New("obj cannot be converted to runtime.Unstructured")
+	}
+
+	if !reflect.ValueOf(u).Elem().IsValid() {
+		return nil, errors.New("obj cannot be nil")
+	}
+
+	value, _, _ := unstructured.NestedFieldNoCopy(u.UnstructuredContent(), "status", "experimentalConditions")
+	if conditions, ok := value.([]interface{}); ok {
+		return convertUnstructuredConditions(conditions), nil
+	}
+
+	value, _, _ = unstructured.NestedFieldNoCopy(u.UnstructuredContent(), "status", "conditions")
+	if conditions, ok := value.([]interface{}); ok {
+		return convertUnstructuredConditions(conditions), nil
+	}
+
+	return nil, errors.New("obj must have Status with one of Conditions or ExperimentalConditions")
+}
+
+func convertUnstructuredConditions(conditions []interface{}) []metav1.Condition {
+	if conditions == nil {
+		return nil
+	}
+
+	convertedConditions := make([]metav1.Condition, 0, len(conditions))
+	for _, c := range conditions {
+		cMap, ok := c.(map[string]interface{})
+		if !ok || cMap == nil {
+			// TODO: think about returning an error in this case and when type of status are not set (as a signal it is a condition type)
+			continue
+		}
+
+		var conditionType string
+		if v, ok := cMap["type"]; ok {
+			conditionType = v.(string)
+		}
+
+		var status string
+		if v, ok := cMap["status"]; ok {
+			status = v.(string)
+		}
+
+		var observedGeneration int64
+		if v, ok := cMap["observedGeneration"]; ok {
+			observedGeneration = v.(int64)
+		}
+
+		var lastTransitionTime metav1.Time
+		if v, ok := cMap["lastTransitionTime"]; ok {
+			_ = lastTransitionTime.UnmarshalQueryParameter(v.(string))
+		}
+
+		var reason string
+		if v, ok := cMap["reason"]; ok {
+			reason = v.(string)
+		}
+
+		var message string
+		if v, ok := cMap["message"]; ok {
+			message = v.(string)
+		}
+
+		convertedConditions = append(convertedConditions, metav1.Condition{
+			Type:               conditionType,
+			Status:             metav1.ConditionStatus(status),
+			ObservedGeneration: observedGeneration,
+			LastTransitionTime: lastTransitionTime,
+			Reason:             reason,
+			Message:            message,
+		})
+	}
+	return convertedConditions
+}
+
+func getFromTypedObject(obj runtime.Object) ([]metav1.Condition, error) {
```

Comment:
Yes, there won't be any GetConditions for metav1.Conditions (all the utils are accepting "plain" runtime.Objects)

Reply:
Yup, but we could check for an interface type with the GetConditions func here instead of using reflection to check for the field. But yeah maybe it's better to stop requiring the Getter